src/boot/efi/console.c

```diff
@@ -70,47 +67,49 @@ EFI_STATUS console_key_read(UINT64 *key, BOOLEAN wait) {
 
 static EFI_STATUS change_mode(UINTN mode) {
         EFI_STATUS err;
+        UINTN old_mode = ST->ConOut->Mode->Mode;
 
         err = uefi_call_wrapper(ST->ConOut->SetMode, 2, ST->ConOut, mode);
+        if (!EFI_ERROR(err))
+                return EFI_SUCCESS;
 
-        /* Special case mode 1: when using OVMF and qemu, setting it returns error
-         * and breaks console output. */
-        if (EFI_ERROR(err) && mode == 1)
-                uefi_call_wrapper(ST->ConOut->SetMode, 2, ST->ConOut, (UINTN)0);
+        /* Something went wront. Output is probably borked, so try to revert to previous mode. */
+        if (!EFI_ERROR(uefi_call_wrapper(ST->ConOut->SetMode, 2, ST->ConOut, old_mode)))
+                return err;
 
+        /* Maybe the device is on fire? */
+        uefi_call_wrapper(ST->ConOut->Reset, 2, ST->ConOut, TRUE);
+        uefi_call_wrapper(ST->ConOut->SetMode, 2, ST->ConOut, (UINTN)0);
         return err;
 }
 
 static UINT64 text_area_from_font_size(void) {
         EFI_STATUS err;
-        UINT64 text_area;
         UINTN rows, columns;
+        const UINT64 SYSTEM_FONT_WIDTH = 8;
+        const UINT64 SYSTEM_FONT_HEIGHT = 19;
 
         err = uefi_call_wrapper(ST->ConOut->QueryMode, 4, ST->ConOut, ST->ConOut->Mode->Mode, &columns, &rows);
         if (EFI_ERROR(err)) {
                 columns = 80;
                 rows = 25;
         }
 
-        text_area = SYSTEM_FONT_WIDTH * SYSTEM_FONT_HEIGHT * (UINT64)rows * (UINT64)columns;
-
-        return text_area;
+        return SYSTEM_FONT_WIDTH * SYSTEM_FONT_HEIGHT * (UINT64)rows * (UINT64)columns;
 }
 
-static EFI_STATUS mode_auto(UINTN *mode) {
-        const UINT32 HORIZONTAL_MAX_OK = 1920;
-        const UINT32 VERTICAL_MAX_OK = 1080;
-        const UINT64 VIEWPORT_RATIO = 10;
-        UINT64 screen_area, text_area;
+static UINTN get_auto_mode(void) {
         EFI_GUID GraphicsOutputProtocolGuid = EFI_GRAPHICS_OUTPUT_PROTOCOL_GUID;
         EFI_GRAPHICS_OUTPUT_PROTOCOL *GraphicsOutput;
-        EFI_GRAPHICS_OUTPUT_MODE_INFORMATION *Info;
         EFI_STATUS err;
         BOOLEAN keep = FALSE;
 
         err = LibLocateProtocol(&GraphicsOutputProtocolGuid, (VOID **)&GraphicsOutput);
         if (!EFI_ERROR(err) && GraphicsOutput->Mode && GraphicsOutput->Mode->Info) {
-                Info = GraphicsOutput->Mode->Info;
+                const UINT32 HORIZONTAL_MAX_OK = 1920;
+                const UINT32 VERTICAL_MAX_OK = 1080;
+                const UINT64 VIEWPORT_RATIO = 10;
```

Comment:
use #defines. We don't use constants for trivial types like this. It makes little sense in C really, it's a C++ thing